### PR TITLE
bpo-38368: Added fix for ctypes crash when handling arrays in structs…

### PR DESCRIPTION
**Pull request by vsajip**
…/unions.

https://bugs.python.org/issue38368

### CODE REVIEW

**Comment by vsajip** [comment]
Tests run on Windows x64 but not on Linux. Still investigating.

File: Modules/_ctypes/stgdict.c
Comment (ZackerySpytz) [inline Modules/_ctypes/stgdict.c]:
```suggestion
                ffi_type *dummy_struct;
```

File: Modules/_ctypes/stgdict.c
Comment (ZackerySpytz) [inline Modules/_ctypes/stgdict.c]:
```suggestion
                ffi_type **dummy_fields;
```

File: Modules/_ctypes/stgdict.c
Comment (ZackerySpytz) [inline Modules/_ctypes/stgdict.c]: I think there should be a `Py_DECREF(pair);` here.

File: Modules/_ctypes/stgdict.c
Comment (ZackerySpytz) [inline Modules/_ctypes/stgdict.c]: I think `dummy_struct` should be freed here. Also, `pair` should be decrefed.

Comment (vsajip) [reply Modules/_ctypes/stgdict.c]: Well, the problem is that we're doing these allocations in a loop, and unless we keep some sort of linked list or dynamically allocated structure to keep track of all our allocations (both `dummy_struct` and `dummy_fields` in earlier iterations of the loop), we will leak this memory in an Out Of Memory (OOM) situation. I'm not aware that we currently do this type of recovery for these internal allocations in an OOM situation, but happy to be proved wrong.

Comment (vsajip) [reply Modules/_ctypes/stgdict.c]: I will have a think about an alternative allocation strategy which allows an easier recovery, but it will be potentially a little harder to follow the logic compared to what's there now.

**Review by zooba — APPROVED** [review]
I'd like to have one more person go over this line-by-line as well to track the memory usage, but I didn't spot anything that seemed wrong.

**Comment by vsajip** [comment]
> I'd like to have one more person go over this line-by-line as well to track the memory usage

Agreed - please request a review from a core developer who you think might have the bandwidth to do this. I asked various people to review the earlier PR (for issue 22273) but didn't get many takers ...

File: Modules/_ctypes/stgdict.c
Comment (ericsnowcurrently) [inline Modules/_ctypes/stgdict.c]: Why not use `Py_DECREF()`?  Don't we already know that `pair` isn't NULL?

File: Modules/_ctypes/stgdict.c
Comment (ericsnowcurrently) [inline Modules/_ctypes/stgdict.c:818]: This check (but certainly not the call) might be redundant based on the earlier for loop.

File: Modules/_ctypes/stgdict.c
Comment (ericsnowcurrently) [inline Modules/_ctypes/stgdict.c:827]: This check seems redundant based on the earlier for loop.

**Review by ericsnowcurrently — APPROVED** [review]
LGTM.  I've left some minor comments, but they aren't critical.

I did double-check the memory allocations and everything looks good.  The only allocations were the 2 `PySequence_GetItem()` calls and the one `PyMem_Malloc()` call.  They are cleaned up in all the right spots.

File: Modules/_ctypes/stgdict.c
Comment (vsajip) [reply Modules/_ctypes/stgdict.c:818]: Although it _shouldn't_ fail, I suppose one can't guarantee it? For that reason, I'd leave the check in, it makes the code a wee bit bigger but shouldn't really impact performance.

File: Modules/_ctypes/stgdict.c
Comment (vsajip) [reply Modules/_ctypes/stgdict.c:827]: As per earlier comment - would prefer to leave the check in for now.

**Comment by vsajip** [comment]
Thanks for the feedback, @zooba and @ericsnowcurrently !

File: Modules/_ctypes/stgdict.c
Comment (vsajip) [reply Modules/_ctypes/stgdict.c]: Good point, I'll change that.

**Comment by vsajip** [comment]
I'll wait for the buildbots to pronounce on this merge before backporting to 3.8/3.7.

**Comment by doko42** [comment]
I checked that at least two packages in Debian and Ubuntu, python-ipaddr and python-pysnmp4 now work again with this patch (only tested on the 3.7 branch).

**Comment by vsajip** [comment]
@doko42 Thanks for the feedback.

**Pull request merged**
